src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -54,10 +67,199 @@ import qualified Data.ByteString.Short.Internal as SBS
 import           Foreign.Ptr
 
+{------------------------------------------------------------------------------
+  TODO: placeholders until #55 is merged
+-------------------------------------------------------------------------------}
+
+data Append =
+    -- | One or more keys are in this page, and their values fit within a single
+    -- page.
+    AppendSinglePage S.SerialisedKey S.SerialisedKey
+    -- | There is only one key in this page, and it's value does not fit within
+    -- a single page.
+  | AppendMultiPage S.SerialisedKey Word32 -- ^ Number of overflow pages
+
+-- | Append a new page entry to a mutable compact index.
+append :: Append -> MCompactIndex s -> ST s [Chunk]
+append = error "append: replace by Database.LSMTree.Internal.Run.Index.Compact.append"
+
+mkAppend :: PageAcc -> Append
+mkAppend p = case unStricterList $ pageKeys p of
+    []                     -> error "mkAppend: empty list"
+    [k] | numBytes <= 4096 -> AppendSinglePage k k
+        | otherwise        -> AppendMultiPage  k (fromIntegral $ numBytes `rem` 4096 - 1)
+    ks                     -> AppendSinglePage (last ks) (head ks)
+  where
+    numBytes = pageSizeNumBytes $ pageSize p
+
+{-------------------------------------------------------------------------------
+  Incremental, in-memory run construction
+-------------------------------------------------------------------------------}
+
+-- | A mutable structure that keeps track of incremental, in-memory run
+-- construction.
+--
+-- Use 'new' to start run construction, add new key\/operation pairs to the run
+-- by using 'add' and co, and complete run construction using 'finalise'.
+data MRun s = MRun {
+      mbloom         :: !(MBloom s RawKey)
+    , mindex         :: !(MCompactIndex s)
+      -- | Used to construct the full CompactIndex when run construction is
+      -- completed.
+    , indexChunksRef :: !(STRef s [[Index.Chunk]]) -- TODO: make stricter
+    , currentPageRef :: !(STRef s PageAcc)
+    }
+
+-- | @'new' npages@ starts an incremental run construction.
+--
+-- @npages@ should be an upper bound on the number of pages that will be yielded
+-- by incremental run construction.
+new :: Int -> ST s (MRun s)
+new npages = do
+    mbloom <- Bloom.newEasy 0.1 npages -- TODO: tune bloom filter
+    let rfprec = Index.suggestRangeFinderPrecision npages
+    mindex <- Index.new rfprec 100 -- TODO(optimise): tune chunk size
+    indexChunksRef <- newSTRef []
+    currentPageRef <- newSTRef paEmpty
+    pure MRun{..}
+
+-- | Finalise an incremental run construction. Do /not/ use an 'MRun' after
+-- finalising it.
+--
+-- The frozen bloom filter and compact index will be returned, along with the
+-- final page of the run (if necessary), and the remaining chunks of the
+-- incrementally constructed compact index.
+finalise ::
+     MRun s
+  -> ST s ( Maybe SizedBuilder
+          , Index.Chunk
+          , Index.FinalChunk
+          , Bloom RawKey
+          , CompactIndex
```

Comment:
This conveniently gives you the full index at the end, but you also get all these index chunks as you go. Is that so you can write out the chunks of the primary array to a file incrementally, but also have the option to just use the full index (without accumulating chunks yourself) e.g. if you construct the run in one go?